packages/@react-stately/color/src/useHexColorFieldState.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {Color} from './Color';
+import {ColorInput, HexColorFieldProps} from '@react-types/color';
+import {NumberFieldState} from '@react-stately/numberfield';
+import {useCallback, useState} from 'react';
+import {useControlledState} from '@react-stately/utils';
+
+export interface HexColorFieldState extends NumberFieldState {
+  colorValue: Color,
+  setInputValue: (value: string) => void
+}
+
+export const defaultMinValue = '#000000';
+export const defaultMaxValue = '#FFFFFF';
+
+export function useHexColorFieldState(
+  props: HexColorFieldProps
+): HexColorFieldState {
+  const {
+    minValue = defaultMinValue,
+    maxValue = defaultMaxValue,
+    step = 1,
+    value,
+    defaultValue,
+    onChange,
+    validationState
+  } = props;
+
+  const minColor = Color.parse(minValue);
+  const maxColor = Color.parse(maxValue);
+  const minColorInt = minColor.toHexInt();
+  const maxColorInt = maxColor.toHexInt();
+
+  const clampColor = useCallback((value: ColorInput) => {
+    try {
+      const color = Color.parse(value);
+      const colorInt = color.toHexInt();
+      if (colorInt < minColorInt) { return minColor; }
+      if (colorInt > maxColorInt) { return maxColor; }
+      return color;
+    } catch (err) {
+      return undefined;
+    }
+  }, [minColorInt, maxColorInt]);
+
+  const initialValue = clampColor(value);
+  const initialDefaultValue = clampColor(defaultValue);
+  const [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
+
+  const initialInputValue = (value || defaultValue) && colorValue ? colorValue.toString('hex') : '';
+  const [inputValue, setInputValue] = useState(initialInputValue);
+
+  const increment = () => {
+    setColorValue((previousValue: Color) => {
+      const colorInt = previousValue ? previousValue.toHexInt() : minColorInt;
```

Comment:
if this colorInt is greater than the max, just return without a new color to prevent causing too many onChange
same in the other functions below
otherwise implement one of the suggestions above so you don't create a new Color

Reply:
Resolved. Also updated `decrement`, `incrementToMax`, and `decrementToMin` to not call `onChange` when the value is already at min/max. Also added another check for when the input changes but the color value is the same ("#CCC <-> #CCCCCC")